cpp/src/spatial/linestring_distance.cu

```diff
@@ -0,0 +1,356 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuspatial/error.hpp>
+#include <cuspatial/types.hpp>
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/utilities/device_atomics.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+#include <limits>
+#include <memory>
+#include <type_traits>
+
+namespace cuspatial {
+namespace {
+
+/** @brief Get the index that is one-past the end point of linestring at @p linestring_idx
+ *
+ * @note The last endpoint of the linestring is not included in the offset array, thus
+ * @p num_points is returned.
+ */
+template <typename OffsetIterator>
+inline cudf::size_type __device__
+endpoint_index_of_linestring(cudf::size_type const& linestring_idx,
+                             OffsetIterator const& linestring_offsets_begin,
+                             cudf::size_type const& num_linestrings,
+                             cudf::size_type const& num_points)
+{
+  return (linestring_idx == (num_linestrings - 1)
+            ? (num_points)
+            : *(linestring_offsets_begin + linestring_idx + 1)) -
+         1;
+}
+
+/**
+ * @brief Computes shortest distance between @p C and segment @p A @p B
+ */
+template <typename T>
+double __device__ point_to_segment_distance(coord_2d<T> const& C,
+                                            coord_2d<T> const& A,
+                                            coord_2d<T> const& B)
+{
+  double L_squared = (A.x - B.x) * (A.x - B.x) + (A.y - B.y) * (A.y - B.y);
+  if (L_squared == 0) { return hypot(C.x - A.x, C.y - A.y); }
+  double r = ((C.x - A.x) * (B.x - A.x) + (C.y - A.y) * (B.y - A.y)) / L_squared;
+  if (r <= 0 or r >= 1) {
+    return std::min(hypot(C.x - A.x, C.y - A.y), hypot(C.x - B.x, C.y - B.y));
+  }
+  double Px = A.x + r * (B.x - A.x);
+  double Py = A.y + r * (B.y - A.y);
+  return hypot(C.x - Px, C.y - Py);
+}
+
+/**
+ * @brief Computes shortest distance between two segments that doesn't intersect.
+ */
+template <typename T>
+double __device__ segment_distance_no_intersect_or_collinear(coord_2d<T> const& A,
+                                                             coord_2d<T> const& B,
+                                                             coord_2d<T> const& C,
+                                                             coord_2d<T> const& D)
+{
+  return std::min(std::min(point_to_segment_distance(A, C, D), point_to_segment_distance(B, C, D)),
+                  std::min(point_to_segment_distance(C, A, B), point_to_segment_distance(D, A, B)));
```

Comment:
This is overkill. The minimum distance corresponds to the minimum squared distance, which is cheaper to compute. So find the minimum squared distance, and then take the square root. This will save at least 3 sqrt/hypot calls.
```suggestion
  auto dist_sqr = std::min(std::min(point_to_segment_distance_squared(A, C, D), point_to_segment_distance_squared(B, C, D)),
                  std::min(point_to_segment_distance_squared(C, A, B), point_to_segment_distance_squared(D, A, B)));
  return sqrt(dist_sqr);
```